src/t_hash.c

```diff
@@ -228,28 +284,27 @@ int hashTypeSet(robj *o, sds field, sds value, int flags) {
         o->ptr = zl;
 
         /* Check if the listpack needs to be converted to a hash table */
-        if (hashTypeLength(o) > server.hash_max_listpack_entries) hashTypeConvert(o, OBJ_ENCODING_HT);
-    } else if (o->encoding == OBJ_ENCODING_HT) {
-        dict *ht = o->ptr;
-        dictEntry *de, *existing;
+        if (hashTypeLength(o) > server.hash_max_listpack_entries) hashTypeConvert(o, OBJ_ENCODING_HASHTABLE);
+    } else if (o->encoding == OBJ_ENCODING_HASHTABLE) {
+        hashtable *ht = o->ptr;
+
         sds v;
         if (flags & HASH_SET_TAKE_VALUE) {
             v = value;
             value = NULL;
         } else {
             v = sdsdup(value);
         }
-        de = dictAddRaw(ht, field, &existing);
-        if (de) {
-            dictSetVal(ht, de, v);
-            if (flags & HASH_SET_TAKE_FIELD) {
-                field = NULL;
-            } else {
-                dictSetKey(ht, de, sdsdup(field));
-            }
+
+        hashtablePosition position;
+        void *existing;
+        if (hashtableFindPositionForInsert(ht, field, &position, &existing)) {
+            /* does not exist yet */
+            hashTypeEntry *entry = hashTypeCreateEntry(field, v);
+            hashtableInsertAtPosition(ht, entry, &position);
         } else {
-            sdsfree(dictGetVal(existing));
-            dictSetVal(ht, existing, v);
+            /* exists: replace value */
+            hashTypeEntryReplaceValue(existing, v);
```

Comment:
I think `hashTypeEntryReplaceValue` should return the entry, so we don't have to change the signature in the future. With embedded value, this function may need to reallocate the entry.

For now, we can just assert that it returns `existing`.

```suggestion
            serverAssert(hashTypeEntryReplaceValue(existing, v) == existing);
```

Reply:
It's only called in one place, and it's a static function - it'll be quite easy to change later. Do you think we'll have some kind of trouble from leaving it simpler like this until embedded values are implemented?

Reply:
OK, leave it. We can refactor it when we need it.